lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -127,7 +127,33 @@ def inline_metadata(catalog, checksum_type)
       file = resource.to_ral
 
       if file.recurse?
-        # TODO: find and replace metadata for children recursively
+        file.recurse_remote_metadata do |meta|
+          # Don't create a new resource for the parent directory
+          next if meta.relative_path == "."
+
+          child_resource = Puppet::Resource.new(:file, File.join(file[:path], meta.relative_path))
+          child_resource[:ensure] = meta.ftype
+          # TODO: Make the coying of owner, group, and mode conditional on using permissions
+          child_resource[:owner]  = meta.owner.to_s
+          child_resource[:group]  = meta.group.to_s
+          child_resource[:mode]   = meta.mode.to_s(8).rjust(4, '0')
+          child_resource[:source] = meta.source
+
+          if meta.ftype == "file"
+            child_resource[:checksum_value] = sumdata(meta.checksum)
+            child_resource[:checksum] = meta.checksum_type
+          end
+
+          # Copy parameters from original parent directory
+          file.original_parameters.each do |param, value|
+            # These should never be passed to our children
+            unless [:parent, :ensure, :recurse, :recurselimit, :target, :alias, :source].include? param
+              child_resource[param] = value.to_s
+            end
+          end
+          catalog.add_resource(child_resource)
+          catalog.add_edge(file, child_resource)
```

Comment:
Short version: We need to add an edge between the `child_resource` and its nearest ancestor (not necessarily the `file` we started recursing in). We also need to add an edge (I think) between the `child_resource` and all other resources that depend on `file`. I think the relationship logic is sufficient complex that we should ticket it separately.

Long version: Currently when an agent sources a recursive directory during [`eval_generate`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/file.rb#L472), it makes a `file_metadatas` request (plural) and for each `metadata` object, constructs a `Puppet::Type::File` instance. It returns the array of generated resources to the [`additional_resource_generator`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/transaction/additional_resource_generator.rb#L54-L58). The generator does 3 things:

1. Adds the generated resources to the catalog
1. Ensures all of the generated resources are contained in the recursive directory resource
1. Adds relationships from the each generated resource to its nearest ancestor.

I'm realizing now that the static compiler has more bugs. It does 1, doesn't handle 2, and incorrectly does 3.

About 2, the static compilation process needs to make sure all of the generated resources come before whatever comes after the file resource we started recursing in, so we don't regress in https://github.com/puppetlabs/puppet/commit/479cd1c087099c481009a68e57b6e0f617b790f3. For example, if you have:

```puppet
file { '/tmp/dir':
  ensure => directory,
  recurse => true,
  source => 'puppet:///modules/mymodule/dir'
} ->
notify { 'after tmp dir': }
```

Then we need to make sure all of the children of `/tmp/dir` are evaluated before the `notify`. The agent currently handles that logic in [`contain_generated_resources_in`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/transaction/additional_resource_generator.rb#L71-L86).

However, we can't create containment edges in the compiler. But I think we can do what the `contain` function does by [adding relationship edges from each generated resource to the `notify`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/parser/functions/contain.rb#L34-L36)

The third issue is that we need to add an edge from the nearest ancestor to the generated resource. So create an edge like:

```
      "source": "File[/tmp/dir/dir1]",
      "target": "File[/tmp/dir/dir1/dir2]"
```

Instead of:

```
      "source": "File[/tmp/dir]",
      "target": "File[/tmp/dir/dir1/dir2]"
```

Otherwise there's no guarantee that `/tmp/dir/dir1` will be created before `/tmp/dir/dir1/dir2`. The agent has similar logic for [eval generated resources](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/transaction/additional_resource_generator.rb#L97-L106). We could also do nothing and rely on autorequiring to create the edges for us.